File: hacking_with_perl/custom_headers.pl
#!/usr/bin/perl
use strict;
use warnings;
use LWP::UserAgent;
use Term::ANSIColor;

# Function to test for custom header vulnerability
sub test_custom_header_vulnerability {
    my $url = shift;

    # Create a UserAgent object
    my $ua = LWP::UserAgent->new;

    # Add custom headers to the request
    $ua->default_header('X-Test-Header' => 'Testing');

    # Send a GET request to the target URL
    my $response = $ua->get($url);

    # Check if the response includes the custom header
    if ($response->header('X-Test-Header')) {
        print colored("The target website is vulnerable to Custom Header Vulnerabilities.\n", 'green');
    } else {
        print colored("The target website is not vulnerable to Custom Header Vulnerabilities.\n", 'red');
    }
}

# Main program
print "Enter your target website URL: ";
my $target_url = <STDIN>;
chomp($target_url);

print "Testing for Custom Header Vulnerabilities on $target_url...\n";

# Call the function to test for vulnerability
test_custom_header_vulnerability($target_url);

# Explanation on how to test the vulnerability
print "\nTo test the vulnerability, you can use tools like curl or Postman to send a request with a custom header, ";
print "such as 'X-Test-Header: Testing', and observe if the header is reflected in the response.\n";

File: hacking_with_ruby/custom_headers.rb
require 'net/http'
require 'colorize'

def test_custom_header_vulnerability(url)
  puts "Testing for Custom Header Vulnerability on #{url}...".yellow

  uri = URI(url)
  response = Net::HTTP.get_response(uri)

  if response['X-Custom-Header']
    puts "Custom header detected: #{response['X-Custom-Header']}".green
    puts "This website may be vulnerable to Custom Header Injection!".red
    puts "To test the vulnerability, try injecting a custom header in your request."
  else
    puts "No custom header detected.".green
    puts "This website does not appear to be vulnerable to Custom Header Injection.".green
  end
end

puts "Enter your target website url: "
target_website = gets.chomp

test_custom_header_vulnerability(target_website)
